compiler/runtime/OMRRelocationRecord.hpp

```diff
@@ -0,0 +1,226 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2022 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef OMR_RELOCATION_RECORD_INCL
+#define OMR_RELOCATION_RECORD_INCL
+/*
+ * The following #define(s) and typedef(s) must appear before any #includes in this file
+ */
+#ifndef OMR_RELOCATION_RECORD_CONNECTOR
+#define OMR_RELOCATION_RECORD_CONNECTOR
+namespace OMR { class RelocationRecord; }
+namespace OMR { typedef OMR::RelocationRecord RelocationRecordConnector; }
+#else
+   #error OMR::RelocationRecord expected to be a primary connector, but another connector is already defined
+#endif
+
+
+#ifndef OMR_RELOCATION_RECORD_BINARY_TEMPLATE_CONNECTOR
+#define OMR_RELOCATION_RECORD_BINARY_TEMPLATE_CONNECTOR
+namespace OMR { class RelocationRecordBinaryTemplate; }
+namespace OMR { typedef OMR::RelocationRecordBinaryTemplate RelocationRecordBinaryTemplateConnector; }
+#else
+   #error OMR::RelocationRecordBinaryTemplate expected to be a primary connector, but another connector is already defined
+#endif
+
+
+#include <stdint.h>
+#include "compile/Compilation.hpp"
+#include "env/jittypes.h"
+#include "infra/Link.hpp"
+#include "infra/Flags.hpp"
+#include "runtime/TRRelocationRuntime.hpp"
+#include "runtime/Runtime.hpp"
+
+namespace TR {
+class RelocationRuntime;
+class RelocationTarget;
+class RelocationRecord;
+class RelocationRecordBinaryTemplate;
+struct RelocationRecordMethodCallPrivateData
+   {
+   uintptr_t callTargetOffset;
+   };
+
+union RelocationRecordPrivateData
```

Comment:
Isn't this union supposed to hold all the associated private data structs, so that when it's included in `RelocationRecord`, sufficient space is allocated to hold the largest of them? I'm not sure how a downstream project is supposed to add new structs here for the new relocation record types defined by that project?

Has the mechanism here changed? I think a comment is warranted to explain how this part is designed to work.

Reply:
Hello Mark, the mechanism has not changed, I have updated the union to properly contain the struct. I think the int that got there was there as an artifact of rebasing. I have added some comments too!